node/network/bitfield-distribution/src/lib.rs

```diff
@@ -0,0 +1,684 @@
+
+		relay_message(ctx, job_data, &mut tracker.peer_views, validator, message).await
+	} else {
+		return modify_reputation(ctx, origin, COST_SIGNATURE_INVALID).await;
+	}
+}
+
+/// Deal with network bridge updates and track what needs to be tracked
+/// which depends on the message type received.
+async fn handle_network_msg<Context>(
+	ctx: &mut Context,
+	tracker: &mut Tracker,
+	bridge_message: NetworkBridgeEvent,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	match bridge_message {
+		NetworkBridgeEvent::PeerConnected(peerid, _role) => {
+			// insert if none already present
+			tracker.peer_views.entry(peerid).or_insert(View::default());
+		}
+		NetworkBridgeEvent::PeerDisconnected(peerid) => {
+			// get rid of superfluous data
+			tracker.peer_views.remove(&peerid);
+		}
+		NetworkBridgeEvent::PeerViewChange(peerid, view) => {
+			catch_up_messages(ctx, tracker, peerid, view).await?;
+		}
+		NetworkBridgeEvent::OurViewChange(view) => {
+			let old_view = std::mem::replace(&mut (tracker.view), view);
+
+			for new in tracker.view.difference(&old_view) {
+				if !tracker.per_relay_parent.contains_key(&new) {
+					warn!(target: "bitd", "Our view contains {} but the overseer never told use we should work on this", &new);
+				}
+			}
+		}
+		NetworkBridgeEvent::PeerMessage(remote, bytes) => {
+			if let Ok(gossiped_bitfield) = BitfieldGossipMessage::decode(&mut (bytes.as_slice())) {
+				trace!(target: "bitd", "Received bitfield gossip from peer {:?}", &remote);
+				process_incoming_peer_message(ctx, tracker, remote, gossiped_bitfield).await?;
+			} else {
+				return modify_reputation(ctx, remote, COST_MESSAGE_NOT_DECODABLE).await;
+			}
+		}
+	}
+	Ok(())
+}
+
+// Send the difference between two views which were not sent
+// to that particular peer.
+async fn catch_up_messages<Context>(
+	ctx: &mut Context,
+	tracker: &mut Tracker,
+	origin: PeerId,
+	view: View,
+) -> SubsystemResult<()>
+where
+	Context: SubsystemContext<Message = BitfieldDistributionMessage>,
+{
+	use std::collections::hash_map::Entry;
+	let current = tracker.peer_views.entry(origin.clone()).or_default();
+
+	let delta_vec: Vec<Hash> = (*current).difference(&view).cloned().collect();
+
+	*current = view;
+
+	// Send all messages we've seen before and the peer is now interested
+	// in to that peer.
+
+	let delta_set: HashMap<ValidatorId, BitfieldGossipMessage> = delta_vec
+		.into_iter()
+		.filter_map(|new_relay_parent_interest| {
+			if let Some(job_data) = (&*tracker).per_relay_parent.get(&new_relay_parent_interest) {
+				// send all messages
+				let one_per_validator = job_data.one_per_validator.clone();
+				let origin = origin.clone();
+				Some(
+					one_per_validator
+						.into_iter()
+						.filter(move |(validator, _message)| {
+							// except for the ones the peer already has
+							// let validator = validator.clone();
+							job_data.message_from_validator_needed_by_peer(&origin, validator)
```

Comment:
I would expect this to be true by definition, except for the fact that we don't check that the stuff peers send us corresponds to their view. So a peer can gossip us a message corresponding to hash H even when H isn't in their view. I believe that is OK, even though it's not something the code here is capable of doing.

Anyway, all this to say, a comment here on why this isn't a tautology would be helpful for the reader.

Reply:
The storage of known messages is per relay parent, so a combination of all valid messages we received or sent indexed by validator. But each of the peers might have already received a subset, and to calculate the subset from the joint known distribution the filter statement is needed.